Increase flexibility of profile with policies implementation.

matejak opened the pull request:
This PR contains just one commit, and is therefore a bit messy, but the majority of changes is just code moving around.

- Restructure profile classes, so their behavior can be extended by inheritance.
- Made ResolvableProfile generic profile that knows the concept of controls, without implementation details.
- Introduced `ProfileWithInlinePolicies`, renamed existing implementation to `ProfileWithSeparatePolicies`.
- Extended the test suite, so it is able to test more implementations in a scalable way.

I expect that things could settle when a concrete profile format is picked.

matejak committed:
Increase flexibility of profile with policies implementation.

- Restructure profile classes, so their behavior can be extended by inheritance.
- Made ResolvableProfile generic profile that knows the concept of controls, without implementation details.
- Introduced ProfileWithInlinePolicies, renamed existing implementation to ProfileWithSeparatePolicies.
- Extended the test suite, so it is able to test more implementations in a scalable way.

matejak commented on ssg/build_yaml.py:
```diff
@@ -111,7 +112,17 @@ def __init__(self, id_):
         self.refine_rules = defaultdict(list)
-        self.policies = {}
+
+    def read_yaml_contents(self, yaml_contents):
```
This has been extracted from `from_yaml`, and it allows extension of yaml processing in derived classes.

matejak commented on ssg/build_yaml.py:
```diff
@@ -166,37 +166,28 @@ def dump_yaml(self, file_name, documentation_complete=True):
 
     def _parse_selections(self, entries):
         for item in entries:
-            if "." in item:
-                rule, refinement = item.split(".", 1)
-                property_, value = refinement.split("=", 1)
-                if property_ not in XCCDF_REFINABLE_PROPERTIES:
-                    msg = ("Property '{property_}' cannot be refined. "
-                           "Rule properties that can be refined are {refinables}. "
-                           "Fix refinement '{rule_id}.{property_}={value}' in profile '{profile}'."
-                           .format(property_=property_, refinables=XCCDF_REFINABLE_PROPERTIES,
-                                   rule_id=rule, value=value, profile=self.id_)
-                           )
-                    raise ValueError(msg)
-                self.refine_rules[rule].append((property_, value))
-            elif "=" in item:
-                varname, value = item.split("=", 1)
-                self.variables[varname] = value
-            elif item.startswith("!"):
-                self.unselected.append(item[1:])
-            else:
-                self.selected.append(item)
```
Extracted to `apply_selection`, again to allow extension of parsing `selections:` in derived classes.

matejak commented on ssg/build_yaml.py:
```diff
@@ -166,37 +166,28 @@ def dump_yaml(self, file_name, documentation_complete=True):
-            else:
-                self.selected.append(item)
-
-    def _parse_policies(self, policies_yaml):
-        for item in policies_yaml:
-            id_ = required_key(item, "id")
-            controls = required_key(item, "controls")
-            if not isinstance(controls, list):
-                if controls != "all":
-                    msg = ("Policy {id_} contains invalid controls list {controls}."
-                        .format(id_=id_, controls=str(controls))
-                        )
-                    raise ValueError(msg)
-            self.policies[id_] = controls
```
This method has been moved to a `ProfileWithSeparatePolicies`, which is called indirectly from `ResolvableProfile`

matejak commented on ssg/build_yaml.py:
```diff
@@ -335,64 +326,141 @@ def __sub__(self, other):
+        control_id_list = controls_manager.get_all_controls(policy_id)
+        return self._controls_to_items(controls_manager, policy_id, control_id_list)
+
+    def resolve_controls(self, controls_manager):
+        pass
```
By default, the profile doesn't know how to resolve controls, although it has all those utility methods. So it does nothing, as the backwards compatibility suggests.

matejak commented on ssg/build_yaml.py:
```diff
@@ -335,64 +326,141 @@ def __sub__(self, other):
+    def resolve_controls(self, controls_manager):
+        pass
+
+    def extend_by(self, extended_profile):
+        extended_selects = set(extended_profile.selected)
+        self.resolved_selections.update(extended_selects)
+
+        updated_variables = dict(extended_profile.variables)
+        updated_variables.update(self.variables)
+        self.variables = updated_variables
+
+        updated_refinements = dict(extended_profile.refine_rules)
+        updated_refinements.update(self.refine_rules)
+        self.refine_rules = updated_refinements
```
`resolve` now does more than just handling the inheritance, so I have extracted the parent profile merge to a separate method.

matejak commented on ssg/build_yaml.py:
```diff
@@ -335,64 +326,141 @@ def __sub__(self, other):
             if self.extends not in all_profiles:
                 msg = (
-                    "Profile {name} extends profile {extended}, but"
+                    "Profile {name} extends profile {extended}, but "
                     "only profiles {known_profiles} are available for resolution."
                     .format(name=self.id_, extended=self.extends,
-                            profiles=list(all_profiles.keys())))
+                            known_profiles=list(all_profiles.keys())))
```
These are just unrelated bugfixes.

matejak commented on tests/unit/ssg-module/test_controls.py:
```diff
@@ -113,6 +139,8 @@ def test_profile_resolution_extends():
     assert high_profile.variables["var_password_pam_ocredit"] == "2"
 
+
+@pytest.mark.skip
```
The `abcd-all.profile` file is not there.

jan-cerny replied:
Why was it removed?

jan-cerny replied:
Aha, I see, It accidentally wasn't committed in #5, I have added a new commit to #5. I'm sorry for confusion. You can re-base and remove it.

jan-cerny commented on ssg/build_yaml.py:
```diff
@@ -335,64 +326,141 @@ def __sub__(self, other):
+        self.resolved = False
+        self.resolved_selections = set()
+
+    def _control_to_items(self, controls_manager, policy_id, control_id):
+        return controls_manager.get_control(policy_id, control_id)
+
```
I think the name is confusing. Also, it doesn't do anything else than calling `get_control`. It's only renaming it.

matejak replied:
OK, I will discard it.

jan-cerny commented on ssg/build_yaml.py:
```diff
@@ -335,64 +326,141 @@ def __sub__(self, other):
 
     def resolve_controls(self, controls_manager):
         for policy_id, controls in self.policies.items():
+            to_expand = Control()
```
Can you explain this? Later, it replaces the controls_list which used to be a list of controls objects by a single instance of a control object.

matejak replied:
I have changed the flow closer to the original one. See the 3rd commit.

jan-cerny commented on ssg/build_yaml.py:
```diff
@@ -335,64 +326,141 @@ def __sub__(self, other):
+
             if isinstance(controls, list):
-                controls_list = [controls_manager.get_control(policy_id, control_id) for control_id in controls]
+                to_expand = self._controls_to_items(controls_manager, policy_id, controls)
```
`_controls_to_items` is defined in `ResolvableProfile`, but not here and not in `Profile`.

matejak replied:
Good point, if those special profile classes build upon `ResolvableProfile`, it will streamline things.

jan-cerny commented:
I think there is something missing in the patch. I think that you probably need to replace `ResolvableProfile` by  `SeparateResolvableProfile` in `build-scripts/compile_profiles.py:43`. Currently I'm getting a traceback during the build if I try to use the controls in a profile in a real product:

```
Traceback (most recent call last):                                                                                                                                                            
  File "/home/jcerny/work/git/scap-security-guide/ssg/build_yaml.py", line 640, in add_profiles_from_dir                                                                                          new_profile = Profile.from_yaml(dir_item_path, env_yaml)                                                                                                                                  
  File "/home/jcerny/work/git/scap-security-guide/ssg/build_yaml.py", line 139, in from_yaml                                                                                                  
    raise RuntimeError("Unparsed YAML data in '%s'.\n\n%s"                                                                                                                                    RuntimeError: Unparsed YAML data in '/home/jcerny/work/git/scap-security-guide/rhel8/./profiles/abcd-low.profile'.                                                                            
                                                                                                                                                                                              
{'policies': [{'id': 'abcd', 'controls': ['R1', 'R2', 'R3']}]}                                                                                                                                                                                                                                                                                                                              
During handling of the above exception, another exception occurred:                                                                                                                           
                                                                                                                                                                                              Traceback (most recent call last):                                                                                                                                                            
  File "/home/jcerny/work/git/scap-security-guide/build-scripts/yaml_to_shorthand.py", line 95, in <module>                                                                                   
    main()                                                                                                                                                                                      File "/home/jcerny/work/git/scap-security-guide/build-scripts/yaml_to_shorthand.py", line 89, in main                                                                                       
    loader.process_directory_tree(benchmark_root, add_content_dirs)                                                                                                                           
  File "/home/jcerny/work/git/scap-security-guide/ssg/build_yaml.py", line 1335, in process_directory_tree                                                                                        self._load_group_process_and_recurse(start_dir)                                                                                                                                           
  File "/home/jcerny/work/git/scap-security-guide/ssg/build_yaml.py", line 1321, in _load_group_process_and_recurse                                                                           
    self.loaded_group = self.load_benchmark_or_group(guide_directory)                                                                                                                           File "/home/jcerny/work/git/scap-security-guide/ssg/build_yaml.py", line 1419, in load_benchmark_or_group                                                                                   
    result = super(ListInputsLoader, self).load_benchmark_or_group(guide_directory)                                                                                                           
  File "/home/jcerny/work/git/scap-security-guide/ssg/build_yaml.py", line 1311, in load_benchmark_or_group                                                                                       group.add_profiles_from_dir(self.action, self.profiles_dir, self.env_yaml)                                                                                                                
  File "/home/jcerny/work/git/scap-security-guide/ssg/build_yaml.py", line 646, in add_profiles_from_dir                                                                                      
    raise RuntimeError(msg)                                                                                                                                                                   RuntimeError: Error building profile from '/home/jcerny/work/git/scap-security-guide/rhel8/./profiles/abcd-low.profile': 'Unparsed YAML data in '/home/jcerny/work/git/scap-security-guide/rhe
l8/./profiles/abcd-low.profile'.                                                                                                                                                              
                                                                                                                                                                                              {'policies': [{'id': 'abcd', 'controls': ['R1', 'R2', 'R3']}]}'  
```

matejak committed:
Improvements

- Added test for the all-inline case.
- Stramlined the inheritance relation of profiles.
- Removed _control_to_items as it was not helpful.

matejak committed:
Streamlined controls processing.

- Prevented conversion from control to ID and to control again in _controls_to_items
- Removed _policy_to_items
- Modified the flow, so now controls are obtained in a list, and then merged to the profile one-by-one.
